### PR TITLE
Deprecate dataset errors with module `__getattr__`

### PR DESCRIPTION
**Pull request by deepyaman**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Deprecating using `DeprecatedClassMeta` only deprecates the name when a deprecated object is instantiated. In the case of errors, they are often used without instantiation (e.g. in `except` blocks), so `DeprecatedClassMeta` doesn't work.

## Development notes

Use `__getattr__` approach for deprecation. This is a relatively-newer method (added in 3.7, see https://peps.python.org/pep-0562/).

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: tests/io/test_core.py
Comment (astrojuanlu) [inline tests/io/test_core.py:39]: Does this work if we use https://docs.python.org/3/library/importlib.html#importlib.import_module instead?

File: kedro/io/core.py
Comment (astrojuanlu) [inline kedro/io/core.py:82]: What about having a mapping between old names new objects so that we don't have to use the `globals()` mechanism? Something like

```python
_DEPRECATED_ERROR_CLASSES = {
    "DataSetError": DatasetError,
    "DataSetNotFoundError": DatasetNotFoundError,
    "DataSetAlreadyExistsError": DatasetAlreadyExistsError,
}
```

the `if name in _DEPRECATED_ERROR_CLASSES` would still work, and then we could do `return _DEPRECATED_ERROR_CLASSES[name]`.

File: kedro/io/__init__.py
Comment (astrojuanlu) [inline kedro/io/__init__.py:37]: While I oppose using `globals()` for the exceptions, this is fine by me 👍🏽

**Review by astrojuanlu — COMMENTED** [review]
Thanks a lot @deepyaman! Left a few suggestions for improvement.

**Review by noklam — COMMENTED** [review]
some minor comments on styling, I have done manual testing and it's working, but I have some questions.


image

Questions:
We will get different a error when we do `from xxx import` versus `importlib.import_module`, is this fine? I think it should be fine, as this behavior is consistent when I test a 3-rd party library like `pandas`, we didn't change anything, but just want to check to make sure.


ref: the code I used for manual test
```python
# Expected

from kedro.io import DataSetNotFoundError

# Expected
import importlib
io = importlib.import_module("kedro.io")
io.DataSetNotFoundError

# ImportError
from kedro.io import DUMMY

# AttributeError
io.DUMMY
```

File: tests/io/test_core.py
Comment (deepyaman) [reply tests/io/test_core.py:39]: No, it doesn't. :( `exec` was not even on my list of things to try (I didn't even know about it?), but nothing else seemed to work.

File: kedro/io/core.py
Comment (deepyaman) [reply kedro/io/core.py:82]: The code with `globals()` pretty much comes from the PEP that proposed adding module `__getattr__`: https://peps.python.org/pep-0562/

But I don't know why it shouldn't work to use a dictionary here, so I can try if `globals()` is bad to use. My understanding was that this may be a good use of `globals()`.

I can also see how other libraries using `__getattr__` for depreciation do it, potentially.

File: tests/io/test_core.py
Comment (astrojuanlu) [reply tests/io/test_core.py:39]: Weird, we did some manual tests and seemed to work https://github.com/kedro-org/kedro/pull/2673#issuecomment-1598916317 in any case, if it doesn't, it's fine

File: kedro/io/__init__.py
Comment (deepyaman) [reply kedro/io/__init__.py:38]: I created https://github.com/kedro-org/kedro/pull/2706 to fix it for most cases. I realized that I did indeed also write the same code again i this PR, so I've also updated it here (and updated the test).

File: kedro/io/core.py
Comment (deepyaman) [reply kedro/io/core.py]: `globals()` removed in https://github.com/kedro-org/kedro/pull/2673/commits/9b13345b52b9c23baea74b7f24aa96493ebfdd57 (following your suggestion).

About the `exec()`, I'm not sure what to do/haven't looked into it more. Open to suggestions; also OK to ignore it for now since it's only in test code that can be removed post-0.19.0-release.

Comment (deepyaman) [reply kedro/io/core.py]: Actually, I will try overwriting `__dir__` and see if it happens to help.

File: tests/io/test_core.py
Comment (deepyaman) [reply tests/io/test_core.py]: Oh, wow, I'm dumb. I tried `importlib.import_module("DataSetError", "kedro.io.core")` and stuff like that, but obviously you can't import a class using `import_module`. Fixing...

**Review by astrojuanlu — APPROVED** [review]
💯

File: kedro/io/__init__.py
Comment (astrojuanlu) [reply kedro/io/__init__.py:38]: gh-2706 is merged!

**Comment by astrojuanlu** [comment]
I don't understand why DCO was happy until now, but then when I updated the branch it started complaining about all the previous commits, and even after I removed my merge commit, it keeps complaining. @deepyaman do you want to do the honors and signoff everything?

**Review by tynandebold — APPROVED** [review]

**Review by noklam — APPROVED** [review]

**Comment by astrojuanlu** [comment]
Submitted a linearized version of this to gh-2711.

**Pull request merged**
